### PR TITLE
persist: implementation of Consensus with postgres

### PR DESCRIPTION
**Pull request by ruchirK**
### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/persist/src/postgres.rs
Comment (danhhz) [inline src/persist/src/postgres.rs:36]: any reason not to just make the internals of this a `tokio_postgres::Config`?

File: src/persist/src/postgres.rs
Comment (danhhz) [inline src/persist/src/postgres.rs:120]: can we use BIGINT to at least make this an i64?

**Review by danhhz — COMMENTED** [review]
sorry for the drive by! feel free to ignore until this is ready to review

File: src/persist/src/postgres.rs
Comment (danhhz) [inline src/persist/src/postgres.rs:42]: we should figure out how to set this in CI sooner than later. probably also fine as a followup but we should track it in  11605 so it doesn't get dropped

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs]: made it a url string which i felt was even simpler

File: src/persist/src/error.rs
Comment (ruchirK) [reply src/persist/src/error.rs]: removed, thanks!

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs]: implemented tokio_postgres's `ToSql` / `FromSql` directly for `SeqNo` to avoid all the type casts.

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs]: vestigial. removed here and elsewhere

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs]: this was a typo. meant to be >= to validate that the returned sequence number didn't overflow i64::MAX. this is all simpler now with the `ToSql` / `FromSql` impls

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs]: done.

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs]: done.

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs:164]: Did a brief search for this -- it seems like our only hope in the driver / postgres itself is something like
`statement_timeout` as defined on this page https://www.postgresql.org/docs/current/runtime-config-client.html

but this feels very unlike our deadline parameter (in that we want per-request deadlines, but this setting is more like a one time setting that we don't want to modify often). Given that it seems like `select!` + sleep is probably the best way to go?

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs]: done.

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs:105]: I'm pretty sure (and pls sanity check me here) that we can't add the code to complain about this in ci without also wiring up some testing in ci as well right?

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs:83]: I think the way I've wired it up worked? I didn't even need to add postgres to the mz composition I just added a new env var. I don't have a clear artifact of whether i correctly wired things up or not other than - before wiring the postgres consensus test succeeded in 0.00008 seconds, and after wiring the env variable through the test took .28 seconds.

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs]: Added this line in + plumbed the env variable through.

File: src/persist/src/postgres.rs
Comment (danhhz) [reply src/persist/src/postgres.rs:83]: i also agree with nikhil. until we hit something that's different between the two for our usage then we should just be testing vs the easy thing

> also sanity check manually that the implementation as written works with aurora.

double checking that you did this and it worked?

File: src/persist/src/postgres.rs
Comment (danhhz) [reply src/persist/src/postgres.rs:169]: > but this feels very unlike our deadline parameter (in that we want per-request deadlines

part of what we're derisking here is things like timeouts. e.g. is our deadline parameter even the best way to model timeouts? so far, it's got two strikes given that it doesn't seem to want to play well with sqlite or postgres

concretely, here I think we need to do the `select!` thing. I poked around the tokio_postgres docs and found `https://docs.rs/tokio-postgres/latest/tokio_postgres/struct.Client.html#method.cancel_token` which seems like something we could in the timeout case to cancel the work itself in postgres (in addition to unblocking in our code). I haven't looked at this deeply enough to know what the performance implications are of using this, whether it works with aurora, etc. that'd be good for you to look into

I'm fine leaving this as a followup to keep performance work/the demo unblocked, but let's add this to my tracking issue before you merge so it doesn't get dropped

**Review by danhhz — COMMENTED** [review]
okay I think this is about ready, modulo a couple more bits of feedback on SeqNo restrictions

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs:165]: Added a comment to do this in a fast follow and will also file in your tracking issue.

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs]: done

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs]: done

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs:37]: :)

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs]: Done. Added a commit that introduces this restriction to all consensus impls, changes sqlite and mem to handle it correctly (involved another From/ToSql impl for sqlite) and aded 4 unit tests to check 0, i64::max, i64::max + 1, and u64::max

File: src/persist/src/location.rs
Comment (danhhz) [inline src/persist/src/location.rs:1067]: this is better expressed as just asserting on the compare_and_set call with things inlined. as-is, the "table" based testing doesn't give you much indication about which case failed. you could get around this by adding more contexts to the asserts but in this case, it's really not too much boilerplate to just assert directly

```rust
assert_eq!(consensus.compare_and_set(&"zero", deadline, None, VersionedData {
    seqno: SeqNo(0),
    data: vec![],
}).await, Ok(()));
```

(or however rustfmt deals with that line)

**Review by danhhz — APPROVED** [review]
lgtm modulo one last thing!

File: src/persist/src/location.rs
Comment (ruchirK) [reply src/persist/src/location.rs:1067]: ok will do!

**Comment by ruchirK** [comment]
@danhhz - I'm double checking that this works with aurora and then assuming that works + the last few nits then I'll merge! TFTR!

File: src/persist/src/postgres.rs
Comment (ruchirK) [reply src/persist/src/postgres.rs:81]: yup! just tested against aurora and it works!

**Comment by ruchirK** [comment]
TFTR! merging on green!

**Pull request merged**
